src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -389,7 +398,8 @@ private void signalIfAllRegisteredInQueue() {
         // is active at a time - a caller of lock() that has just acquired
         // the pool lock but is waiting for all registered elements to be
         // returned to the queue.
-        if (registeredElements.size() == liveQueue.size()) {
+        if (registeredElements.size() == liveQueue.size() ||
+                poisonPill != null) {
```

Comment:
I was thinking that throwing the exception would take care of getting stuck waiting for the lock, but you're right, if the pill is inserted after the await, then it'll deadlock. 

Well at least the name's explicit